### DISCUSSION TITLE
Typer expects default value for CustomClass as string instead of class itself, which breaks typing and breaks using this function without typer

### DISCUSSION THREAD

**Discussion by AlexeyDmitriev** [fastapi/typer#720] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer
import datetime
from typing_extensions import Annotated


def parse(s: str):
    return datetime.datetime.strptime(s, "%Y-%m-%d").date()

def main(date: Annotated[datetime.date, typer.Option(parser=parse)] = "2023-12-18"):
    print(date)

if __name__ == '__main__':
    typer.run(main)
```

### Description

So example code above works fine, but, it doesn't type check properly because main supposed to accept datetime.date, and it's default is string.
Also, if I try to run function main without typer(a big feature for typer that it's usually possible), this argument will get str

If I change it to 

```python
import typer
import datetime
from typing_extensions import Annotated


def parse(s: str):
    return datetime.datetime.strptime(s, "%Y-%m-%d").date()

def main(date: Annotated[datetime.date, typer.Option(parser=parse)] = datetime.date(2023, 12, 18)):
    print(date)

if __name__ == '__main__':
    typer.run(main)
```

which type-checks fine, I'll get `TypeError: strptime() argument 1 must be str, not datetime.date` because typer expects that I give it a string, not datetime and calls parser on it

### Operating System

Linux, Windows, macOS, Other

### Operating System Details

any

### Typer Version

0.9.0

### Python Version

Python 3.11.4

### Additional Context

I think I'm ready to (try to) contribute this as PR, if this is considered worth fixing

**Comment by ambroisie** [comment]
I was recently bitten by a similar issue, providing `default_factory=MyClass` and `parser=MyClass` (where `MyClass` can be built without arguments, or with one string positional argument) leads to `typer` trying to call `MyClass(MyClass())` (the outer one is the parser, the inner one the default value).

**Comment by svlandeg** [comment]
I can see how this construction is preventing a proper type check, as the `parser` is being applied to the parameter value, whether it is user-defined or default, which means the default needs to be a `str` and not a `CustomClass` 🤔 

> I think I'm ready to (try to) contribute this as PR, if this is considered worth fixing

How would you attempt to fix this?

**Reply by AlexeyDmitriev** [reply]
> How would you attempt to fix this?

So I see 3 main ways to fix it
1) Instead of FuncParamType https://github.com/pallets/click/blob/8.1.x/src/click/types.py implement another ParamType that 
will first check if it's given str or the type and if it's a type then return the same value, without trying to parse it.

I'm pretty sure this will work, even if it feels a bit dirty.

2) To look how exactly typer and click interact here and ensure that in case of default, click doesn't interact here
For example, one possible way is to have optional string click parameter and postprocess yourself only if needed (if nothing is passed, return default otherwise call parser(given_string_value)

I'm not sure it's doable without complete rewrite of how typer works (not deep enough in this code)

3) Rewrite how click uses FuncParamType so that this requirement is not needed for them
https://github.com/pallets/click/blob/8.1.x/src/click/types.py#L34 

Also not sure how feasible is that.

If I'd make a PR, I'd do the first way.
